### ISSUE TITLE
Migrate to semantic logging by default?

### ISSUE THREAD

**Issue by rogeralsing** [akkadotnet/akka.net#1045]
The internal logging in Akka.NET is based on the standard approach of generating strings of text that are written to a log.
Pretty much `String.Format("My Template {0} - Other stuff {1}",arg1,arg2)`

This works fine when logging to loggers such as NLog. but it works less ideal when working with a semantic logger like Serilog, as you will end up with a lot of data bound to the term "{0}" where all of that data represents totally different things.

We can not extract meaning from this data in retrospect, we can only look at the lines and read them out as text.

The old approach is not compatible with semantic logging, but semantic logging is compatible with the old approach.

If we did have an abstraction ontop of our logs, that lets us write the above exampe as
`log.Info("Remoting is listening at port {PortNumber}",portNumber)`

We could then use this data to extract realtime'ish statistics. portnumber might be a lame case, but lets take cluster events instead, it would be pretty awesome to see how often your nodes disconnect/rejoin the cluster.

I'm not sure how this could be accomplished, but it sure would be a nice feature that adds to the monitoring story.

**Comment by stefansedich** [comment]
I like this! I am camping so cant respond much but I will reply next week
with my thoughts. Having done some work to get serilog to work nicely
previosuly I have some thoughts here.

Cheers
Stefan

**Comment by Horusiath** [comment]
1. What about performance?
2. Could we support both?

**Comment by rogeralsing** [comment]
One way to solve it for the internals, would to make the actual log message into a resource that can be swapped out.

e.g.

`log.Info("foo bar {0}",yadda);`
becomes
`log.Info(LogStrings.FooBar,yadda);`

Then each individual logger could provide their own set of templates for all the internal messages.
That would be a beast to maintain through. any new log message would have to be added to each logger.

**Comment by stefansedich** [comment]
@rogeralsing I agree that having to have logging strings per logger would be a nightmare to maintain.

I wonder if we move to semantic based logging by default, we could then remove the need for the Default/Serilog message formatters that I added a while ago.

We can then look at migrating to our own semantic based message formatter (or even bring in the one from Serilog which does a good job) for the logging tools that do not support it we simply format the message before logging.

@Horusiath as far as performance goes I did not see a huge increase when using the serilog formatter over say string.Format.


Thanks
Stefan

**Comment by nblumhardt** [comment]
One way I've seen this work is to use a structured format for the internal implementation, e.g.:

```csharp
log.Info("Foo is {Bar}", yadda);
```

which is passed directly to structured loggers like Serilog that support it. (The new ASP.NET 5 logging pipeline is also compatible with this format).

When a text logger is plugged in, either:

* the formatting is performed ahead of passing the string to the text logger, so essentially `"Foo is 123"` gets passed through, or
* the message template and arguments are used to construct an object of a type like `StructuredLogMessage` that is passed to the text logger as an object, on which `ToString()` is eventually called and the format rendered out.

This has the advantage of keeping the overhead of writing log messages low (using resource dictionaries could be a bit of effort to keep in sync over time).

HTH - if there's anything I can add please let me know (I'm one of the maintainers of Serilog).

Cheers!

**Comment by rogeralsing** [comment]
@nblumhardt what would be the best approach to re-use the serilog template parser?
copy paste reuse from the Serilog code into the Akka logging infrastructure?
I'd like to avoid to take dependency directly to Serilog from Akka core if possible.

**Comment by rogeralsing** [comment]
IMO, the current API where we can use either standard index based format, _or_ semantic names in the same methods is not a good approach imo.

if some bit of code gets passed a logger instance. there is no way to determine what kind of format that logger expects.
pass in the wrong format and the logger might blow up.

also, by changing the akka config, the loggers might blow up as the new loggers might not play nicely with your existing formatting.

My conclusion here is that our logger API is a mess, I think we should have a distinction between index based and semantic logging methods.

**Comment by rogeralsing** [comment]
I've written a PoC that transforms standard index based format tempates into semantic tempates.

e.g.
`Drained buffer with maxWriteCount: {0}, fullBackoffCount: {1},smallBackoffCount: {2}, noBackoffCount: {3},adaptiveBackoff: {4}`

Becomes

`Drained buffer with maxWriteCount: {DrainedBufferWithMaxWriteCount}, fullBackoffCount: {FullBackoffCount},smallBackoffCount: {SmallBackoffCount}, noBackoffCount: {NoBackoffCount},adaptiveBackoff: {AdaptiveBackoff}`

This is ofcourse not as good as if you had named the properties yourself, but IMO it's close enough for extracting some value out it if a semantic logger is used.

**Comment by stefansedich** [comment]
How does it handle things that are not prefixed with anything of value
"Starting {0}:{1}" where for example 0 is Host and 1 is Port, I worry that
if we don't get this right now from the get go that we will again end up
with sub-standard logging that we are not happy with.



Cheers
Stefan

**Comment by rogeralsing** [comment]
@stefansedich I absolutely agree, and my vote here goes to change our code to use true semantic logging.
But I suspect that even if we all agree on this, it will take a while before we are actually there.

my PoC could be used in the Serilog adapter, or in a fork of it. to allow users to extract some value out of our existing logs.
So it was not intended as a final solution but rather an experiment to see if the current logs could be used for data extraction.

**Comment by nblumhardt** [comment]
@rogeralsing regarding the parser - yes, I think `Ctrl+C`, `Ctrl+V` would be the recommended approach - keeps dependencies etc. down :-)

**Comment by rogeralsing** [comment]
I have been playing some more with this.
The built in message types `Debug` `Info` `Warning` `Error` (not the methods, the actual messages) does not have any formatting capabilities, they just contain a dumb string that is then used in the `LogEvent` message base type to create a `ToString` representation.

For Semantic logging inside akka core to be truly useful, that would need to be redesigned.

But it's still pretty neat to see that we can extract information like Heartbeats or disconnects to different systems and display that in Kibana.

I really do think that going all in semantic inside Akka core is the best way forward. 
There is a ton of useful information that can be extracted there.

**Comment by rogeralsing** [comment]
From Roland:

> This sounds like a very sensible thing to do, we are looking into this for the whole stack, but nothing has emerged yet. We’ll definitely switch that on for Akka as soon as we know the shape of that switch ;-)

**Comment by nblumhardt** [comment]
I read that as _"when is someone going to port Serilog to Java?"_ ;-)

(Seriously though, I'd love to give this a shot if someone with some better Java chops was interested. I've considered going it alone, but it's such a rich platform that for an outsider it'd be impossible to make this feel "idiomatic".)
